File: src/main/java/io/github/hydrazinemc/oxidiser/event/world/ExplosionDetonatedEvent.java
package io.github.hydrazinemc.oxidiser.event.world;

import net.minecraft.world.explosion.Explosion;
import io.github.hydrazinemc.oxidiser.event.StimulusEvent;

/**
 * Called when an {@link Explosion} is detonated in the world.
 * <p>
 * This event can be used to modify the blocks affected by the explosion, for example,
 * by modifying {@link Explosion#getAffectedBlocks()}
 */
public interface ExplosionDetonatedEvent {
    StimulusEvent<ExplosionDetonatedEvent> EVENT = StimulusEvent.create(ExplosionDetonatedEvent.class, ctx -> (explosion, particles) -> {
        try {
            for (var listener : ctx.getListeners()) {
                listener.onExplosionDetonated(explosion, particles);
            }
        } catch (Throwable t) {
            ctx.handleException(t);
        }
    });

    void onExplosionDetonated(Explosion explosion, boolean particles);
}

File: src/main/kotlin/io/github/hydrazinemc/oxidiser/event/world/ExplosionDetonatedEvent.kt
package io.github.hydrazinemc.oxidiser.event.world

import io.github.hydrazinemc.oxidiser.event.Event
import net.minecraft.world.explosion.Explosion

/**
 * Called when an [Explosion] is detonated in the world.
 * This event can be used to modify the blocks affected by the explosion, for example,
 * by modifying [Explosion.affectedBlocks]
 */
object ExplosionDetonatedEvent : Event<ExplosionDetonatedEvent.EventData>() {
	data class EventData(val explosion: Explosion, val particles: Boolean)
}